File: backend/app/services/refund_service.py
"""
Graduation Refund Service — list candidates with positive wallet balance and
issue a payout (CASH / BANK_TRANSFER / CHEQUE) that debits the wallet inside a
single transaction.

Note: `is_graduated` is NOT enforced here — cashier verifies visually. We only
require `customer.is_active` and `wallet.is_active` with `balance > 0`.
"""
from __future__ import annotations

from decimal import Decimal
from typing import List, Optional

from sqlalchemy import asc, desc, nulls_last
from sqlalchemy.orm import Session, joinedload

from app.models.customer import Customer
from app.models.wallet import Wallet, WalletTransaction, WalletTransactionType
from app.schemas.refund import RefundCandidate, RefundResponse


_VALID_METHODS = {"CASH", "BANK_TRANSFER", "CHEQUE"}
_METHOD_LABEL = {
    "CASH": "Cash",
    "BANK_TRANSFER": "Bank transfer",
    "CHEQUE": "Cheque",
}


class RefundService:

    # ── Candidate listing ─────────────────────────────────────────────────────

    @staticmethod
    def list_candidates(db: Session) -> List[RefundCandidate]:
        """Return all customers with wallet balance > 0.

        Filter: wallet.balance > 0 AND wallet.is_active AND customer.is_active.
        Sort: is_graduated DESC, withdraw_date DESC NULLS LAST, name ASC.
        """
        rows = (
            db.query(Customer, Wallet)
            .join(Wallet, Wallet.customer_id == Customer.id)
            .filter(
                Wallet.balance > 0,
                Wallet.is_active.is_(True),
                Customer.is_active.is_(True),
            )
            .order_by(
                desc(Customer.is_graduated),
                nulls_last(desc(Customer.withdraw_date)),
                asc(Customer.name),
            )
            .all()
        )
        return [
            RefundCandidate(
                id=customer.id,
                name=customer.name,
                student_code=customer.student_code,
                family_code=customer.family_code,
                is_graduated=customer.is_graduated,
                wallet_id=wallet.id,
                wallet_balance=Decimal(str(wallet.balance)),
                enroll_date=customer.enroll_date,
                withdraw_date=customer.withdraw_date,
            )
            for customer, wallet in rows
        ]

    # ── Refund issuance ───────────────────────────────────────────────────────

    @staticmethod
    def create_refund(
        db: Session,
        customer_id: int,
        amount: Decimal,
        method: str,
        notes: Optional[str],
        user_id: int,
    ) -> RefundResponse:
        """Issue a graduation refund (wallet debit) under one DB transaction.

        Stores `amount` as a positive Decimal — direction is encoded in
        `balance_before` / `balance_after`, matching the convention used by
        `WalletService.adjust_balance` and `ReturnsService` REFUND rows.
        """
        # ── Validation (pre-transaction) ──────────────────────────────────
        if amount is None or Decimal(str(amount)) <= 0:
            raise ValueError("Refund amount must be positive")
        if method not in _VALID_METHODS:
            raise ValueError(
                f"Invalid refund method '{method}'. Must be one of: "
                f"{sorted(_VALID_METHODS)}"
            )
        amount_dec = Decimal(str(amount))

        try:
            # Lock the wallet row to prevent concurrent balance mutation.
            wallet = (
                db.query(Wallet)
                .with_for_update()
                .filter(Wallet.customer_id == customer_id)
                .first()
            )
            if not wallet:
                raise ValueError(f"Wallet not found for customer {customer_id}")
            if not wallet.is_active:
                raise ValueError(f"Wallet {wallet.id} is inactive")

            balance_before = Decimal(str(wallet.balance))
            if amount_dec > balance_before:
                raise ValueError(
                    f"Refund amount ฿{amount_dec:.2f} exceeds wallet balance "
                    f"฿{balance_before:.2f}"
                )

            wallet.balance = balance_before - amount_dec

            note_part = f" — {notes.strip()}" if notes and notes.strip() else ""
            description = (
                f"Graduation refund ({_METHOD_LABEL[method]}){note_part}"
            )

            tx = WalletTransaction(
                wallet_id=wallet.id,
                transaction_type=WalletTransactionType.REFUND,
                amount=amount_dec,
                balance_before=balance_before,
                balance_after=wallet.balance,
                reference_type="graduation_refund",
                reference_id=None,
                description=description,
                reason="graduation_refund",
                refund_method=method,
                reference_ticket=None,
                created_by=user_id,
            )
            db.add(tx)
            db.commit()
            db.refresh(tx)

            return RefundResponse(
                transaction_id=tx.id,
                customer_id=customer_id,
                wallet_id=wallet.id,
                amount=amount_dec,
                refund_method=method,
                balance_before=balance_before,
                balance_after=Decimal(str(wallet.balance)),
                reason="graduation_refund",
                notes=notes,
                created_at=tx.created_at,
                created_by_user_id=user_id,
            )
        except Exception:
            db.rollback()
            raise

File: backend-bun/src/services/refund_service.ts
import { and, asc, desc, eq, gt, sql, sqlNull } from "drizzle-orm";
import { db, pgClient } from "@/db/client";
import { customers, wallets } from "@/db/schema";
import { pgNumber, pgToIso } from "@/lib/dates";

const VALID_METHODS = new Set(["CASH", "BANK_TRANSFER", "CHEQUE"]);
const METHOD_LABEL: Record<string, string> = {
  CASH: "Cash",
  BANK_TRANSFER: "Bank transfer",
  CHEQUE: "Cheque",
};

export interface RefundCandidateDTO {
  id: number;
  name: string;
  student_code: string | null;
  family_code: string | null;
  is_graduated: boolean;
  wallet_id: number;
  wallet_balance: number;
  enroll_date: string | null;
  withdraw_date: string | null;
}

export interface RefundResponseDTO {
  transaction_id: number;
  customer_id: number;
  wallet_id: number;
  amount: number;
  refund_method: string;
  balance_before: number;
  balance_after: number;
  reason: "graduation_refund";
  notes: string | null;
  created_at: string;
  created_by_user_id: number;
}

export async function listRefundCandidates(): Promise<RefundCandidateDTO[]> {
  const rows = await db
    .select({ customer: customers, wallet: wallets })
    .from(customers)
    .innerJoin(wallets, eq(wallets.customerId, customers.id))
    .where(
      and(
        sql`${wallets.balance} > 0`,
        eq(wallets.isActive, true),
        eq(customers.isActive, true),
      ),
    )
    .orderBy(
      desc(customers.isGraduated),
      // NULLS LAST for withdraw_date desc
      sql`${customers.withdrawDate} DESC NULLS LAST`,
      asc(customers.name),
    );

  return rows.map(({ customer, wallet }) => ({
    id: customer.id,
    name: customer.name,
    student_code: customer.studentCode ?? null,
    family_code: customer.familyCode ?? null,
    is_graduated: customer.isGraduated,
    wallet_id: wallet.id,
    wallet_balance: pgNumber(wallet.balance) ?? 0,
    enroll_date: customer.enrollDate ?? null,
    withdraw_date: customer.withdrawDate ?? null,
  }));
}

export async function createGraduationRefund(args: {
  customerId: number;
  amount: number;
  method: string;
  notes: string | null;
  userId: number;
}): Promise<RefundResponseDTO> {
  const { customerId, amount, method, notes, userId } = args;
  if (!Number.isFinite(amount) || amount <= 0) {
    const err = new Error("Refund amount must be positive");
    (err as { status?: number }).status = 400;
    throw err;
  }
  if (!VALID_METHODS.has(method)) {
    const err = new Error(`Invalid refund method '${method}'. Must be one of: CASH, BANK_TRANSFER, CHEQUE`);
    (err as { status?: number }).status = 400;
    throw err;
  }

  return await pgClient.begin(async (sqlTx) => {
    const wRows = await sqlTx<Array<{ id: number; balance: string; is_active: boolean }>>`
      SELECT id, balance, is_active FROM wallets WHERE customer_id = ${customerId} FOR UPDATE
    `;
    if (wRows.length === 0) {
      const err = new Error(`Wallet not found for customer ${customerId}`);
      (err as { status?: number }).status = 404;
      throw err;
    }
    const wallet = wRows[0];
    if (!wallet.is_active) {
      const err = new Error(`Wallet ${wallet.id} is inactive`);
      (err as { status?: number }).status = 400;
      throw err;
    }
    const balanceBefore = Number(wallet.balance);
    if (amount > balanceBefore) {
      const err = new Error(`Refund amount ฿${amount.toFixed(2)} exceeds wallet balance ฿${balanceBefore.toFixed(2)}`);
      (err as { status?: number }).status = 400;
      throw err;
    }
    const balanceAfter = balanceBefore - amount;
    await sqlTx`UPDATE wallets SET balance = ${balanceAfter}, updated_at = NOW() WHERE id = ${wallet.id}`;

    const notePart = notes && notes.trim() ? ` — ${notes.trim()}` : "";
    const description = `Graduation refund (${METHOD_LABEL[method]})${notePart}`;

    const txRows = await sqlTx<Array<{ id: number; created_at: string }>>`
      INSERT INTO wallet_transactions
        (wallet_id, transaction_type, amount, balance_before, balance_after,
         reference_type, reference_id, description, reason, refund_method, created_by)
      VALUES (${wallet.id}, 'REFUND', ${amount}, ${balanceBefore}, ${balanceAfter},
              'graduation_refund', NULL, ${description}, 'graduation_refund',
              ${method}, ${userId})
      RETURNING id, created_at
    `;

    return {
      transaction_id: txRows[0].id,
      customer_id: customerId,
      wallet_id: wallet.id,
      amount,
      refund_method: method,
      balance_before: balanceBefore,
      balance_after: balanceAfter,
      reason: "graduation_refund",
      notes,
      created_at: pgToIso(txRows[0].created_at)!,
      created_by_user_id: userId,
    };
  });
}
